Sessions refactoring (#17649)
symfony/symfony-docs: session.rst

.. index::
   single: HTTP
   single: HttpFoundation, Sessions

Sessions
========

> Installation
> ------------

Reviewer: ```suggestion
```

(just a suggestion, I never like articles starting with a sub-title)

Author: Refactored to avoid that.

The Symfony HttpFoundation component has a very powerful and flexible session
subsystem which is designed to provide session management that you can use to
store information about the user between requests through a clear
object-oriented interface using a variety of session storage drivers.

> You need to install HttpFoundation to handle sessions:

Reviewer: ```suggestion
You need to install the HttpFoundation component to handle sessions:
```

.. code-block:: terminal

    $ composer require symfony/http-foundation

Symfony sessions are designed to replace the usage of the ``$_SESSION`` super
global and native PHP functions related to manipulating the session like
``session_start()``, ``session_regenerate_id()``, ``session_id()``,
``session_name()``, and ``session_destroy()``.

.. note::

    Sessions are only started if you read or write from it.

.. _session-intro:

Basic Usage
-----------

The session is available through the ``Request`` object and the ``RequestStack``
service. Symfony injects the ``request_stack`` service in services and controllers
if you type-hint an argument with :class:`Symfony\\Component\\HttpFoundation\\RequestStack`::

.. configuration-block::

>     .. code-block:: php-symfony

Reviewer 2: Do we have this code-block? 👀

Reviewer: yes

        use Symfony\Component\HttpFoundation\RequestStack;

        class SomeService
        {
            private $requestStack;

            public function __construct(RequestStack $requestStack)
            {
                $this->requestStack = $requestStack;

>                 // Accessing the session in the constructor is *NOT* recommended, since

Reviewer 2: Imho this should be a note and not a code example if it is not recommended

Author: That's something that I copy/pasted, so not new. I would keep it like this as this is about basic usage, and this one is bad.

                // it might not be accessible yet or lead to unwanted side-effects
                // $this->session = $requestStack->getSession();
            }

            public function someMethod()
            {
                $session = $this->requestStack->getSession();

                // ...
            }
        }

    .. code-block:: php-standalone

        use Symfony\Component\HttpFoundation\Session\Attribute\AttributeBag;
        use Symfony\Component\HttpFoundation\Session\Session;
        use Symfony\Component\HttpFoundation\Session\Storage\NativeSessionStorage;

        $session = new Session();
        $session->start();

        // Pass a storage and a bag implementation explicitely
        $session = new Session(new NativeSessionStorage(), new AttributeBag());

From a Symfony controller, type-hint an argument with
:class:`Symfony\\Component\\HttpFoundation\\RequestStack` or
:class:`Symfony\\Component\\HttpFoundation\\Request`::

    use Symfony\Component\HttpFoundation\Request;
    use Symfony\Component\HttpFoundation\RequestStack;
    use Symfony\Component\HttpFoundation\Response;

>     public function index(RequestStack $requestStack): Response

Reviewer 3: Showing two different ways to access the session in the same code snippet might be confusing to, as it is not clear why there are two ways and when to use which.

Maybe we can add a note that Controllers commonly get the session from the `Request` whereas services might want to use the `RequestStack` instead, so that you don't have to pass around the request and/or session between services? Alternatively, it might make sense to update the examples to have one controller- and one service-example showing both ways independently.

Author: Done. I have remove the RequestStack example for the controller as this is never something you need.
As we already have an example for a service just above, I think it's good now.

    {
        $session = $requestStack->getSession();

        // ...
    }

    public function index(Request $request): Response
    {
        $session = $request->getSession();

        // ...
    }

> Attributes
> ----------

Reviewer 3: Only a minor nitpick, but it might mislead people jumping to this section, thinking this is related to PHP 8 Attributes
```suggestion
Session Attributes
------------------
```

Author: Right, fixed

> PHP's session management requires the use of the ``$_SESSION`` super-global,
> however, this interferes somewhat with code testability and encapsulation in an
> OOP paradigm. To help overcome this, Symfony uses *session bags* linked to the

Reviewer: ```suggestion
PHP's session management requires the use of the ``$_SESSION`` super-global.
However, this interferes with code testability and encapsulation in an OOP
paradigm. To help overcome this, Symfony uses *session bags* linked to the
```

        {% for message in app.flashes('notice') %}
            <div class="flash-notice">
                {{ message }}
            </div>
        {% endfor %}

        {# read and display several types of flash messages #}
        {% for label, messages in app.flashes(['success', 'warning']) %}
            {% for message in messages %}
                <div class="flash-{{ label }}">
                    {{ message }}
                </div>
            {% endfor %}
        {% endfor %}

        {# read and display all flash messages #}
        {% for label, messages in app.flashes %}
            {% for message in messages %}
                <div class="flash-{{ label }}">
                    {{ message }}
                </div>
            {% endfor %}
        {% endfor %}

    .. code-block:: php-standalone

        // display warnings
        foreach ($session->getFlashBag()->get('warning', []) as $message) {
            echo '<div class="flash-warning">'.$message.'</div>';
        }

        // display errors
        foreach ($session->getFlashBag()->get('error', []) as $message) {
            echo '<div class="flash-error">'.$message.'</div>';
        }

>         // Display all flashes at once

Reviewer 2: ```suggestion
        // display all flashes at once
```
Lowercase like above

        foreach ($session->getFlashBag()->all() as $type => $messages) {
            foreach ($messages as $message) {
                echo '<div class="flash-'.$type.'">'.$message.'</div>';
            }
        }

It's common to use ``notice``, ``warning`` and ``error`` as the keys of the
different types of flash messages, but you can use any key that fits your
needs.

.. tip::

    You can use the
    :method:`Symfony\\Component\\HttpFoundation\\Session\\Flash\\FlashBagInterface::peek`
    method instead to retrieve the message while keeping it in the bag.

Configuration
-------------

In the Symfony framework, sessions are enabled by default. Session storage and
other configuration can be controlled under the :ref:`framework.session
configuration <config-framework-session>` in
``config/packages/framework.yaml``:

> .. configuration-block::

Reviewer: Let's add a `.. code-block:: php-standalone` here as well (we do that in the HttpClient docs too), applies to other configuration examples as well.

Author: Done. I've done it for some other sections as well, but not for the DB configurations.

logged in by destroying the session after a certain period of idle time. For
example, it is common for banking applications to log the user out after just
5 to 10 minutes of inactivity. Setting the cookie lifetime here is not
appropriate because that can be manipulated by the client, so we must do the expiry
on the server side. The easiest way is to implement this via garbage collection
which runs reasonably frequently. The ``cookie_lifetime`` would be set to a
relatively high value, and the garbage collection ``gc_maxlifetime`` would be set
to destroy sessions at whatever the desired idle period is.

The other option is specifically check if a session has expired after the
session is started. The session can be destroyed as required. This method of
processing can allow the expiry of sessions to be integrated into the user
experience, for example, by displaying a message.

Symfony records some metadata about each session to give you fine control over
the security settings::

    $session->getMetadataBag()->getCreated();
    $session->getMetadataBag()->getLastUsed();

Both methods return a Unix timestamp (relative to the server).

> This metadata can be used to explicitly expire a session on access, e.g.::

Reviewer 2: ```suggestion
This metadata can be used to explicitly expire a session on access::
```

        $session->invalidate();
        throw new SessionExpired(); // redirect to expired session page
    }

It is also possible to tell what the ``cookie_lifetime`` was set to for a
particular cookie by reading the ``getLifetime()`` method::

    $session->getMetadataBag()->getLifetime();

The expiry time of the cookie can be determined by adding the created
timestamp and the lifetime.

.. index::
    single: Session; Database Storage

.. _session-database:

Store Sessions in a Database
----------------------------

Symfony stores sessions in files by default. If your application is served by
multiple servers, you'll need to use a database instead to make sessions work
across different servers.

Symfony can store sessions in all kinds of databases (relational, NoSQL and
key-value) but recommends key-value databases like Redis to get best
performance.

Store Sessions in a key-value Database (Redis)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This section assumes that you have a fully-working Redis server and have also
installed and configured the `phpredis extension`_.

You have two different options to use Redis to store sessions:

> (1) The first PHP-based option is to configure Redis session handler directly
>     in the server ``php.ini`` file:

Reviewer: ```suggestion
The first PHP-based option is to configure Redis session handler directly
in the server ``php.ini`` file:
```

rst tries hard to transform this into a list, breaking the rendering a bit.

.. code-block:: ini

    ; php.ini
    session.save_handler = redis
    session.save_path = "tcp://192.168.0.178:6379?auth=REDIS_PASSWORD"

> (2) The second Symfony-based option is to configure Redis sessions as follows.
> 
> First, define a Symfony service for the connection to the Redis server:

Reviewer: ```suggestion
The second option is to configure Redis sessions in Symfony. First, define
a Symfony service for the connection to the Redis server:
```

Reviewer: slightly related: it looks like we're missing a high-level API for this in the framework.

Is this intended (e.g. do we actually not recommend storing sessions in a db at all) or is this a missing DX feature?

                    <call method="auth">
                        <argument>%env(REDIS_USER)%</argument>
                        <argument>%env(REDIS_PASSWORD)%</argument>
                    </call> -->
                </service>
            </services>
        </container>

    .. code-block:: php

        // ...
        $container
            // you can also use \RedisArray, \RedisCluster or \Predis\Client classes
            ->register('Redis', \Redis::class)
            ->addMethodCall('connect', ['%env(REDIS_HOST)%', '%env(int:REDIS_PORT)%'])
            // uncomment the following if your Redis server requires a password:
            // ->addMethodCall('auth', ['%env(REDIS_PASSWORD)%'])
            // uncomment the following if your Redis server requires a user and a password (when user is not default):
            // ->addMethodCall('auth', ['%env(REDIS_USER)%', '%env(REDIS_PASSWORD)%'])
        ;

Now pass this ``\Redis`` connection as an argument of the service associated to the
:class:`Symfony\\Component\\HttpFoundation\\Session\\Storage\\Handler\\RedisSessionHandler`.
This argument can also be a ``\RedisArray``, ``\RedisCluster``, ``\Predis\Client``,

> and ``RedisProxy``:

Reviewer: I suggest combining the previous and next code snippets into one, and move this to a code comment.

Author: Done

.. configuration-block::

    .. code-block:: yaml

        # config/services.yaml
        services:
            # ...
            Symfony\Component\HttpFoundation\Session\Storage\Handler\RedisSessionHandler:
                arguments:
                    - '@Redis'
                    # you can optionally pass an array of options. The only options are 'prefix' and 'ttl',
                    # which define the prefix to use for the keys to avoid collision on the Redis server
                    # and the expiration time for any given entry (in seconds), defaults are 'sf_s' and null:
                    # - { 'prefix': 'my_prefix', 'ttl': 600 }

    .. code-block:: xml

        <!-- config/services.xml -->

>         <services>
>             <service id="Symfony\Component\HttpFoundation\Session\Storage\Handler\RedisSessionHandler">
>                 <argument type="service" id="Redis"/>
>                 <!-- you can optionally pass an array of options. The only options are 'prefix' and 'ttl',
>                      which define the prefix to use for the keys to avoid collision on the Redis server
>                      and the expiration time for any given entry (in seconds), defaults are 'sf_s' and null:
>                 <argument type="collection">
>                     <argument key="prefix">my_prefix</argument>
>                     <argument key="ttl">600</argument>
>                 </argument> -->
>             </service>
>         </services>

Reviewer: ```suggestion
        <!-- config/services.xml -->
        <?xml version="1.0" encoding="UTF-8" ?>
        <container xmlns="http://symfony.com/schema/dic/services"
            xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
            xsi:schemaLocation="http://symfony.com/schema/dic/services
                https://symfony.com/schema/dic/services/services-1.0.xsd">
                
            <services>
                <service id="Symfony\Component\HttpFoundation\Session\Storage\Handler\RedisSessionHandler">
                    <argument type="service" id="Redis"/>
                    <!-- you can optionally pass an array of options. The only options are 'prefix' and 'ttl',
                         which define the prefix to use for the keys to avoid collision on the Redis server
                         and the expiration time for any given entry (in seconds), defaults are 'sf_s' and null:
                    <argument type="collection">
                        <argument key="prefix">my_prefix</argument>
                        <argument key="ttl">600</argument>
                    </argument> -->
                </service>
            </services>
        </container>
```

        // config/services.php
        use Symfony\Component\DependencyInjection\Reference;
        use Symfony\Component\HttpFoundation\Session\Storage\Handler\RedisSessionHandler;

        $container
            ->register(RedisSessionHandler::class)
            ->addArgument(
                new Reference('Redis'),
                // you can optionally pass an array of options. The only options are 'prefix' and 'ttl',
                // which define the prefix to use for the keys to avoid collision on the Redis server
                // and the expiration time for any given entry (in seconds), defaults are 'sf_s' and null:
                // ['prefix' => 'my_prefix', 'ttl' => 600],
            );

Next, use the :ref:`handler_id <config-framework-session-handler-id>`
configuration option to tell Symfony to use this service as the session handler:

.. configuration-block::

    .. code-block:: yaml

        # config/packages/framework.yaml
        framework:
            # ...
            session:
                handler_id: Symfony\Component\HttpFoundation\Session\Storage\Handler\RedisSessionHandler

    .. code-block:: xml

        <!-- config/packages/framework.xml -->

>         <framework:config>
>             <!-- ... -->
>             <framework:session handler-id="Symfony\Component\HttpFoundation\Session\Storage\Handler\RedisSessionHandler"/>
>         </framework:config>

Reviewer: ```suggestion
        <?xml version="1.0" encoding="UTF-8" ?>
        <container xmlns="http://symfony.com/schema/dic/services"
            xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
            xmlns:framework="http://symfony.com/schema/dic/symfony"
            xsi:schemaLocation="http://symfony.com/schema/dic/services
                https://symfony.com/schema/dic/services/services-1.0.xsd
                http://symfony.com/schema/dic/symfony
                https://symfony.com/schema/dic/symfony/symfony-1.0.xsd">

            <framework:config>
                <!-- ... -->
                <framework:session handler-id="Symfony\Component\HttpFoundation\Session\Storage\Handler\RedisSessionHandler"/>
            </framework:config>
        </container>
```

    .. code-block:: php

        // config/packages/framework.php
        use Symfony\Component\HttpFoundation\Session\Storage\Handler\RedisSessionHandler;
        use Symfony\Config\FrameworkConfig;

        return static function (FrameworkConfig $framework) {
            // ...
            $framework->session()
                ->handlerId(RedisSessionHandler::class)
            ;
        };

Symfony will now use your Redis server to read and write the session data. The
main drawback of this solution is that Redis does not perform session locking,
so you can face *race conditions* when accessing sessions. For example, you may
see an *"Invalid CSRF token"* error because two requests were made in parallel
and only the first one stored the CSRF token in the session.

.. seealso::

    If you use Memcached instead of Redis, follow a similar approach but
    replace ``RedisSessionHandler`` by
    :class:`Symfony\\Component\\HttpFoundation\\Session\\Storage\\Handler\\MemcachedSessionHandler`.

> Store Sessions in a Relational Database (MariaDB, MySQL, PostgreSQL)

Reviewer: ```suggestion
.. _session-database-pdo:

Store Sessions in a Relational Database (MariaDB, MySQL, PostgreSQL)
```

        // config/services.php
        namespace Symfony\Component\DependencyInjection\Loader\Configurator;

        use Symfony\Component\HttpFoundation\Session\Storage\Handler\PdoSessionHandler;

        return static function (ContainerConfigurator $container) {

>             $services = $configurator->services();

Reviewer 4: The variable is called `$container` the line above.
I would change this one, but I know there may still be discussions about this.
(this error seems present in some other blocks)

            // ...
            $framework->session()
                ->handlerId(PdoSessionHandler::class)
            ;
        };

Configuring the Session Table and Column Names
..............................................

The table used to store sessions is called ``sessions`` by default and defines
certain column names. You can configure these values with the second argument
passed to the ``PdoSessionHandler`` service:

.. configuration-block::

    .. code-block:: yaml

        # config/services.yaml
        services:
            # ...

            Symfony\Component\HttpFoundation\Session\Storage\Handler\PdoSessionHandler:
                arguments:
                    - '%env(DATABASE_URL)%'
                    - { db_table: 'customer_session', db_id_col: 'guid' }

    .. code-block:: xml

        <!-- config/services.xml -->
        <?xml version="1.0" encoding="UTF-8" ?>
        <container xmlns="http://symfony.com/schema/dic/services"
            xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
            xsi:schemaLocation="http://symfony.com/schema/dic/services
                https://symfony.com/schema/dic/services/services-1.0.xsd">

            <services>

>                 <service id="Symfony\Component\HttpFoundation\Session\Storage\Handler\PdoSessionHandler" public="false">

Reviewer 4: `public="false"` could be removed as this is the default since Symfony 4.

                ->args([
                    env('DATABASE_URL'),
                    // you can also use PDO configuration, but requires passing two arguments:
                    // 'mysql:dbname=mydatabase; host=myhost; port=myport',
                    // ['db_username' => 'myuser', 'db_password' => 'mypassword'],
                ])
            ;
        };

.. tip::

    When using MySQL as the database, the DSN defined in ``DATABASE_URL`` can
    contain the ``charset`` and ``unix_socket`` options as query string parameters.

    .. versionadded:: 5.3

        The support for ``charset`` and ``unix_socket`` options was introduced
        in Symfony 5.3.

Next, use the :ref:`handler_id <config-framework-session-handler-id>`
configuration option to tell Symfony to use this service as the session handler:

.. configuration-block::

    .. code-block:: yaml

        # config/packages/framework.yaml
        framework:
            session:
                # ...
                handler_id: Symfony\Component\HttpFoundation\Session\Storage\Handler\PdoSessionHandler

    .. code-block:: xml

        <!-- config/packages/framework.xml -->

>         <framework:config>
>             <!-- ... -->
>             <framework:session
>                 handler-id="Symfony\Component\HttpFoundation\Session\Storage\Handler\PdoSessionHandler"/>
>         </framework:config>

Reviewer: ```suggestion
        <?xml version="1.0" encoding="UTF-8" ?>
        <container xmlns="http://symfony.com/schema/dic/services"
            xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
            xmlns:framework="http://symfony.com/schema/dic/symfony"
            xsi:schemaLocation="http://symfony.com/schema/dic/services
                https://symfony.com/schema/dic/services/services-1.0.xsd
                http://symfony.com/schema/dic/symfony
                https://symfony.com/schema/dic/symfony/symfony-1.0.xsd">

            <framework:config>
                <!-- ... -->
                <framework:session
                    handler-id="Symfony\Component\HttpFoundation\Session\Storage\Handler\PdoSessionHandler"/>
            </framework:config>
        </container>
```

        `sess_data` BLOB NOT NULL,
        `sess_lifetime` INTEGER UNSIGNED NOT NULL,
        `sess_time` INTEGER UNSIGNED NOT NULL,
        INDEX `sessions_sess_lifetime_idx` (`sess_lifetime`)
    ) COLLATE utf8mb4_bin, ENGINE = InnoDB;

.. note::

    A ``BLOB`` column type (which is the one used by default by ``createTable()``)
    stores up to 64 kb. If the user session data exceeds this, an exception may
    be thrown or their session will be silently reset. Consider using a ``MEDIUMBLOB``
    if you need more space.

PostgreSQL
++++++++++

.. code-block:: sql

    CREATE TABLE sessions (
        sess_id VARCHAR(128) NOT NULL PRIMARY KEY,
        sess_data BYTEA NOT NULL,
        sess_lifetime INTEGER NOT NULL,
        sess_time INTEGER NOT NULL
    );
    CREATE INDEX sessions_sess_lifetime_idx ON sessions (sess_lifetime);

Microsoft SQL Server
++++++++++++++++++++

.. code-block:: sql

    CREATE TABLE sessions (
        sess_id VARCHAR(128) NOT NULL PRIMARY KEY,
        sess_data NVARCHAR(MAX) NOT NULL,
        sess_lifetime INTEGER NOT NULL,
        sess_time INTEGER NOT NULL,
        INDEX sessions_sess_lifetime_idx (sess_lifetime)
    );

> Store Sessions in a NoSQL Database (MongoDB)

Reviewer: ```suggestion
.. _session-database-mongodb:

Store Sessions in a NoSQL Database (MongoDB)
```

                </service>
            </services>
        </container>

    .. code-block:: php

        // config/services.php
        namespace Symfony\Component\DependencyInjection\Loader\Configurator;

        use Symfony\Component\HttpFoundation\Session\Storage\Handler\MongoDbSessionHandler;

        return static function (ContainerConfigurator $container) {
            $services = $configurator->services();

            $services->set(MongoDbSessionHandler::class)
                ->args([
                    service('doctrine_mongodb.odm.default_connection'),
                ])
            ;
        };

Next, use the :ref:`handler_id <config-framework-session-handler-id>`
configuration option to tell Symfony to use this service as the session handler:

.. configuration-block::

    .. code-block:: yaml

        # config/packages/framework.yaml
        framework:
            session:
                # ...
                handler_id: Symfony\Component\HttpFoundation\Session\Storage\Handler\MongoDbSessionHandler

    .. code-block:: xml

        <!-- config/packages/framework.xml -->

>         <framework:config>
>             <!-- ... -->
>             <framework:session
>                 handler-id="Symfony\Component\HttpFoundation\Session\Storage\Handler\MongoDbSessionHandler"/>
>         </framework:config>

Reviewer: ```suggestion
        <?xml version="1.0" encoding="UTF-8" ?>
        <container xmlns="http://symfony.com/schema/dic/services"
            xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
            xmlns:framework="http://symfony.com/schema/dic/symfony"
            xsi:schemaLocation="http://symfony.com/schema/dic/services
                https://symfony.com/schema/dic/services/services-1.0.xsd
                http://symfony.com/schema/dic/symfony
                https://symfony.com/schema/dic/symfony/symfony-1.0.xsd">

            <framework:config>
                <!-- ... -->
                <framework:session
                    handler-id="Symfony\Component\HttpFoundation\Session\Storage\Handler\MongoDbSessionHandler"/>
            </framework:config>
        </container>
```

     * login. This can be used by the LocaleSubscriber afterwards.
     */
    class UserLocaleSubscriber implements EventSubscriberInterface
    {
        private $requestStack;

        public function __construct(RequestStack $requestStack)
        {
            $this->requestStack = $requestStack;
        }

        public function onInteractiveLogin(InteractiveLoginEvent $event)
        {
            $user = $event->getAuthenticationToken()->getUser();

            if (null !== $user->getLocale()) {
                $this->requestStack->getSession()->set('_locale', $user->getLocale());
            }
        }

        public static function getSubscribedEvents()
        {
            return [
                SecurityEvents::INTERACTIVE_LOGIN => 'onInteractiveLogin',
            ];
        }
    }

.. caution::

    In order to update the language immediately after a user has changed their
    language preferences, you also need to update the session when you change
    the ``User`` entity.

.. index::
   single: Sessions, Session Proxy, Proxy

Session Proxies
---------------

The session proxy mechanism has a variety of uses and this article demonstrates

> two common uses. Rather than using the regular session handler, you can create

Reviewer 3: ```suggestion
two common ones. Rather than using the regular session handler, you can create
```
Avoids using "uses" twice in the same sentence.
